frontend/packages/console-plugin-sdk/src/useExtensions.ts

```diff
@@ -0,0 +1,75 @@
+import { useMemo } from 'react';
+// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+// @ts-ignore: FIXME missing exports due to out-of-sync @types/react-redux version
+import { useSelector, shallowEqual } from 'react-redux';
+import { RootState } from '@console/internal/redux';
+import { FlagsObject } from '@console/internal/reducers/features';
+import { pluginStore } from '@console/internal/plugins';
+import { getFlagsFromState, isExtensionInUse } from './store';
+import { Extension, ExtensionTypeGuard } from './typings';
+
+/**
+ * React hook for consuming Console extensions.
+ *
+ * This hook takes an extension type guard as an argument and returns a list of
+ * extension instances (narrowed according to the type guard) which are currently
+ * in use.
+ *
+ * An extension is in use when
+ *
+ * - it is an always-on extension, i.e. not gated by feature flags
+ * - all feature flags referenced by its `flags` object are resolved to the right
+ *   values
+ *
+ * For example:
+ *
+ * ```ts
+ * import { isFooNavItem, isNavItem } from '@console/plugin-sdk';
+ *
+ * const Example = () => {
+ *   // filter by single extension type
+ *   // return type is FooNavItem[]
+ *   const fooNavItems = useExtensions(isFooNavItem);
+ *
+ *   // filter by union of extension types
+ *   // return type is NavItem[] = (FooNavItem | BarNavItem)[]
+ *   const allNavItems = useExtensions(isNavItem);
+ *
+ *   // process extensions and render component
+ * };
+ * ```
+ *
+ * You may call `useExtensions` multiple times in a single functional component
+ * to retrieve Console extensions of different types. Internally, this hook calls
+ * `useSelector` from `react-redux` library, i.e. each call creates an individual
+ * subscription to the Redux store.
+ *
+ * _For now, the runtime list of extensions is assumed to be immutable._
+ *
+ * @param typeGuard Type guard used to narrow the extension type.
+ */
+export const useExtensions: UseExtensions = (typeGuard) => {
+  const allExtensions = pluginStore.getAllExtensions();
+
+  // Select extensions according to type guard
+  const desiredExtensions = useMemo(() => allExtensions.filter(typeGuard), [
+    allExtensions,
+    typeGuard,
+  ]);
+
+  // Compute flags relevant for gating desired extensions
+  const desiredFlags: FlagsObject = useSelector(
+    (state: RootState) => getFlagsFromState(desiredExtensions, state),
```

Comment:
`getFlagsFromState` will run on every store update, maybe we could consider using `reselect` and comparing if store substate - `state.FLAGS` changed before `getFlagsFromState`

Reply:
From `react-redux` [`useSelector`](https://react-redux.js.org/next/api/hooks#useselector) docs:

> The selector will be run whenever the function component renders. `useSelector()` will also subscribe to the Redux store, and run your selector whenever an action is dispatched.

> With `useSelector()`, returning a new object every time will always force a re-render by default. If you want to retrieve multiple values from the store, you can:
> - Call `useSelector()` multiple times, with each call returning a single field value
> - Use Reselect or a similar library to create a memoized selector that returns multiple values in one object, but only returns a new object when one of the values has changed.
> - Use the `shallowEqual` function from React-Redux as the `equalityFn` argument to `useSelector()`

In our `useExtensions` hook, we're using the 3rd option listed above:

```ts
// Compute flags relevant for gating desired extensions
const desiredFlags: FlagsObject = useSelector<RootState, FlagsObject>(
  (state: RootState) => getFlagsFromState(desiredExtensions, state),
  shallowEqual,
);
```

The `useSelector` hook already memoizes the object selected from Redux store, based on the second argument (`equalityFn`) which is `===` by default, but we're customizing it to `shallowEqual` comparator.

This is similar to `connectToFlags` HOC using `{ areStatePropsEqual: _.isEqual }` with the difference that `_.isEqual` does deep comparison while `shallowEqual` only compares the object's own properties.

Since we're selecting a simple object `{ [key: string]: boolean }` it doesn't really matter if we use `_.isEqual` or `shallowEqual` for comparison.

Reply:
@rawagner But I'm also OK with using `_.isEqual` instead of `shallowEqual` (better parity with `connectToFlags` in terms of comparator), up to you.

Reply:
What I meant is that we could use `reselect` to compare state *before* running `getFlagsFromState`. The second argument in `useSelector` compares result *after* `getFlagsFromState`

Reply:
We could compare `state.FLAGS` with old `state.FLAGS` as `FLAGS` are not updated that often

Reply:
@rawagner Sorry, I misunderstood your comment.

I'll use `reselect` to memoize on `RootState.FLAGS` sub-state as suggested.